### PR TITLE
Support for cleveref-style refs

### PR DESCRIPTION
**Pull request by jonsterling**
I have added support for `cref` and `pref` tag, that produce links `Definition 1.4*a [000X]` and `(Definition 1.5*a [000X])` respectively.

I also refactored the code that produces the long-names of nodes. My code may need some further changes in order to match the Ruby idiom, so please feel free to tear it apart. My motivations were in part to address some of the oddities caused by #38.

### CODE REVIEW

**Comment by jonsterling** [comment]
Ah, it looks like I didn't manage to update the support for collecting referrers.

**Comment by paolobrasolin** [comment]
> Ah, it looks like I didn't manage to update the support for collecting referrers.

What you did in https://github.com/jonsterling/jekyll-sheafy/commit/8d71f5d2656abf01fc2281d362ced7d2c63fac7d#diff-cef05919610621f5fe9f9413512331b7512c762501a4bec0166feeb73da820e6R4 is definitely aligned with an idea I already had: making the regexp used to scan for references configurable in `_config.yml`. Once I close paolobrasolin/jekyll-sheafy#4 you can get back to using the original gem. :+1:

File: _plugins/to_numbering_filter.rb [hunk 1]
```diff
@@ -1,15 +1,19 @@
+def clicks_to_numbering(clicks)
```
Comment (paolobrasolin) [inline _plugins/to_numbering_filter.rb:1]: I'd keep this as a class method of `ToNumberingFilter` to avoid polluting the global namespace.

**Comment by jonsterling** [comment]
OK great! Based on paolobrasolin/jekyll-sheafy#4, it is clear that the functionality in this repo can easily be adjusted once that's complete, so I am merging this branch so I can continue to make progress on writing 😄

File: _plugins/ref_tag.rb [hunk 2]
```diff
@@ -1,6 +1,5 @@
 # frozen_string_literal: true
```
Comment (paolobrasolin) [inline _plugins/ref_tag.rb:1]: Maybe these three classes can be refactored into inheriting base functionality from a single one. :thinking:

File: _plugins/node_info.rb [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+    taxon.nil? ? title : (display_index + (has_title? ? ". #{title}" : ""))
+  end
+
+  def display_title_parenthetical
+    taxon.nil? ? title : (display_index + (has_title? ? " (#{title})" : ""))
+  end
+
+  def aria_label
+    label = taxon.nil? ? title : ("#{genus} #{clicks_to_numbering @node['clicks']}." + (has_title? ? " #{title}" : ""))
+    CGI.escapeHTML label
```
Comment (paolobrasolin) [inline _plugins/node_info.rb:48]: I _believe_ Jekyll has some util function for this, but I'm not sure off the top of my head.

**Review by paolobrasolin — APPROVED** [review]

File: _plugins/to_numbering_filter.rb (hunk 1 above)
Comment (jonsterling) [reply _plugins/to_numbering_filter.rb:1]: Can you tell me how to do that properly? I kept it in there but I couldn't figure out how to access it from outside. I saw some incantation that involved `extend self` that solved my problem, but because I didn't know what that means I was a little scared to do it 😆

**Comment by paolobrasolin** [comment]
> I also refactored the code that produces the long-names of nodes. My code may need some further changes in order to match the Ruby idiom, so please feel free to tear it apart.

@jonsterling I left a few comments around, but everything seems to be ok!
Definitely safe to merge and be _maybe_ be refactored after some battle testing. :smile:

File: _plugins/to_numbering_filter.rb (hunk 1 above)
Comment (paolobrasolin) [reply _plugins/to_numbering_filter.rb:1]: TL;DR: the trick is remembering that inside any object
* `extend X` mixes in methods of a module X as *class* methods,
* `include X` mixes in methods of a module X as *instance* methods.

A module is itself an object (precisely, an instance of the `Module` class) so you can do that to modules too.

Consider this:

```ruby
module Moo
  extend self
  def foo; end
end
```

You're mixing in **all** the methods of the module `self` (which is `Moo`) into `Moo` as class methods, therefore making `foo` available as `Moo.foo`.

You can also be more specific:

```ruby
module Moo
  def foo; end
  module_function :foo
end
```

Or you can be **way more explicit**, _wink wink_, if you want someone else to understand what the heck is going on:

```ruby
module ToNumberingFilter
  def to_numbering(clicks)
    self.class.clicks_to_numbering(clicks)
  end
  def self.clicks_to_numbering(clicks)
    # ...
  end
end
# => ToNumberingFilter.clicks_to_numbering can now be used everywhere
```

Ruby is less mysterious once you realize they're really not joking about the "everything is an object" motto. Of course that's a loaded gun. :laughing:

Comment (jonsterling) [reply _plugins/to_numbering_filter.rb:1]: Thanks, that was a helpful explanation! I've addressed this now in https://github.com/jonsterling/math/commit/463a78671bb07ea89a2735d8d0e8dec1877b193d

Comment (paolobrasolin) [reply _plugins/to_numbering_filter.rb:1]: By the way, the reason why in

```ruby
module Moo
  def met; end
end
```

the method `:met` isn't publicly accessible as `Moo.met` is that `:met` is an _instance_ method of `Moo`, which is an instance of class `Module` but also a singleton object which cannot really be instanced (that's why it acts basically like a namespace).

In short, it's an instance method of an object you can't instance.

File: _plugins/node_info.rb (hunk 3 above)
Comment (paolobrasolin) [reply _plugins/node_info.rb:48]: Nevermind, it's buried a bit too deep to reach: https://github.com/jekyll/jekyll/blob/4fbbefeb7eecff17d877f14ee15cbf8b87a52a6e/lib/jekyll/filters.rb#L66-L108

File: _plugins/to_numbering_filter.rb (hunk 1 above)
Comment (jonsterling) [reply _plugins/to_numbering_filter.rb:1]: Hmm, actually I noticed that the `module_function :foo` has a very different behavior in this case from `extend self`. Not sure exactly what happened, but I will switch to the other.

File: _plugins/ref_tag.rb (hunk 2 above)
Comment (paolobrasolin) [reply _plugins/ref_tag.rb:1]: Will be addressed by #45 :+1:

File: _plugins/to_numbering_filter.rb (hunk 1 above)
Comment (paolobrasolin) [reply _plugins/to_numbering_filter.rb:1]: Huh. Weird! There might be some subtle interference w/ Liquid (see "loaded gun" above).
Let's just use `extend` if it works as you expect. :shrug:
